chain/block_filterer.go

```diff
@@ -0,0 +1,171 @@
+	// external addresses found in a single block.
+	FoundExternal map[waddrmgr.KeyScope]map[uint32]struct{}
+
+	// FoundInternal is a two-layer map recording the scope and index of
+	// internal addresses found in a single block.
+	FoundInternal map[waddrmgr.KeyScope]map[uint32]struct{}
+
+	// RelevantTxns records the transactions found in a particular block
+	// that contained matches from an address in either ExReverseFilter or
+	// InReverseFilter.
+	RelevantTxns []*wire.MsgTx
+}
+
+// NewBlockFilterer constructs the reverse indexes for the current set of
+// external and internal addresses that we are searching for, and is used to
+// scan successive blocks for addresses of interest. A particular block filter
+// can be reused until the first call from `FitlerBlock` returns true.
+func NewBlockFilterer(params *chaincfg.Params, externalAddrs,
+	internalAddrs map[waddrmgr.ScopedIndex]btcutil.Address) *BlockFilterer {
+
+	// Construct a reverse index by address string for the requested
+	// external addresses.
+	nExAddrs := len(externalAddrs)
+	exReverseFilter := make(map[string]waddrmgr.ScopedIndex, nExAddrs)
+	for scopedIndex, addr := range externalAddrs {
+		exReverseFilter[addr.EncodeAddress()] = scopedIndex
+	}
+
+	// Construct a reverse index by address string for the requested
+	// internal addresses.
+	nInAddrs := len(internalAddrs)
+	inReverseFilter := make(map[string]waddrmgr.ScopedIndex, nInAddrs)
+	for scopedIndex, addr := range internalAddrs {
+		inReverseFilter[addr.EncodeAddress()] = scopedIndex
+	}
+
+	foundExternal := make(map[waddrmgr.KeyScope]map[uint32]struct{})
+	foundInternal := make(map[waddrmgr.KeyScope]map[uint32]struct{})
+
+	return &BlockFilterer{
+		Params:          params,
+		ExReverseFilter: exReverseFilter,
+		InReverseFilter: inReverseFilter,
+		FoundExternal:   foundExternal,
+		FoundInternal:   foundInternal,
+	}
+}
+
+// FilterBlock parses all txns in the provided block, searching for any that
+// contain addresses of interest in either the external or internal reverse
+// filters. This method return true iff the block contains a non-zero number of
+// addresses of interest.
+func (bf *BlockFilterer) FilterBlock(block *wire.MsgBlock) bool {
+	var foundAddrs bool
+	for _, tx := range block.Transactions {
+		if bf.FilterTx(tx) {
+			bf.RelevantTxns = append(bf.RelevantTxns, tx)
+			foundAddrs = true
+		}
+	}
+
+	return foundAddrs
+}
+
+// FilterTx scans all txouts in the provided txn, testing to see if any found
+// addresses match those contained within the external or internal reverse
+// indexes. This method returns true iff the txn contains a non-zero number of
+// addresses of interest, indicating that this is a "relevant" txn.
+func (bf *BlockFilterer) FilterTx(tx *wire.MsgTx) bool {
+	var isRelevant bool
+	for i, out := range tx.TxOut {
+		_, addrs, _, err := txscript.ExtractPkScriptAddrs(
+			out.PkScript, bf.Params,
+		)
+		if err != nil {
+			log.Errorf("Could not parse output script in %s:%d: %v",
```

Comment:
Thinking we maybe shouldn't log here, there are a decent number of scripts that don't parse properly